CPS-0006? | Governance Security (#491)
cardano-foundation/CIPs: CPS-?/README.md

---
CPS: ???
Title: Cardano Governance Security

> Category: ?

Reviewer: Fair enough, thanks for the insight @JaredCorduan.

I am unsure which category fits the best, perhaps Meta?

cc: @rphair @KtorZ

> Category: ?

Reviewer 2: yes @Ryun1 @JaredCorduan I would vote for `Meta` especially due to the wide range of areas each of these proposal components touch upon.

> Category: ?

Author: Can I add @Liberty-Chris  @rphair @Ryun1 @JaredCorduan as contributors or authors to this CPS? Not sure which one or what the industry standards are for CPS contributors, but I would like to credit people for their efforts.

> Category: ?

Reviewer 2: @rickymac68 I think that a Contributors section would be welcome but I may not have enough influence on the content to be listed as an Author (generally so for other editors).

There are likely to be a number of people who you want to credit as contributors by the time this CPS makes it through the process but we don't have an official section for it.  The best I can think of is to add an H3 element (`### Contributors`) after the Abstract where you can list those you you feel are key contributors.

Status: Open
Authors:
    - Richard McCracken <rickymac68@icloud.com>
    - Andrew Westberg <andrewwestberg@gmail.com>
Proposed Solutions: N/A
Discussions:
    - https://github.com/cardano-foundation/CIPs/pull/490
    - https://github.com/cardano-foundation/CIPs/pull/491
Created: 2023-03-28
---

# Cardano Governance Security

## Abstract

> As Cardano's community-led governance system goes live, it's crucial to address the security of the governance mechanism in a way that informs all participants of potential weaknesses and how to mitigate them. The guidelines provided in the CIP are relevant to various groups, including Constitutional Committee members, Delegation Representatives, Stake Pool Operators, voters, and all Ada holders. Moreover, the guidelines are applicable to wallet, dApp, and infrastructure developers to help them build situational awareness and design governance security considerations into their development processes.

Reviewer: ```suggestion
As Cardano's community-led governance system goes live, it's crucial to address the security of the governance mechanism in a way that informs all participants of potential weaknesses and how to mitigate them. The guidelines provided in this CPS are relevant to various groups, including Constitutional Committee members, Delegation Representatives, Stake Pool Operators, voters, and all Ada holders. Moreover, the guidelines are applicable to wallet, dApp, and infrastructure developers to help them build situational awareness and design governance security considerations into their development processes.
```

Author: thanks, good fix

In rare instances, a threat may be identified that requires responsible reporting procedures to prevent exposing an active vulnerability to bad actors seeking to exploit it. In such cases, it is essential to notify the appropriate group or entity with the ressources, or are responsible for mitigating the threat in a confidential manner.

### #1: Unknown software vulnerability affecting governance is discovered.

An assumption is that are no known software back doors. Unknown software vulnerabilities may be discovered.

Recommendation: The person discovering the vulnerability should document the circumstances and details in which the vulnerability is present and how to reproduce the vulnerability. Then report the vulnerability in a confidential manner using a secure method to the developer team with the resources to correct the vulnerability.

> 

Reviewer 3: A tiered bug bounty rewards program can be employed to recompense those who articulate meaningful corrections required to reverse/repair a potential vulnerability. A treasury may need to be formed to fund this specific activity, releasing those proposed rewards by a committee vote.

Author: Ok Chris, I will add that to the recommendations. Be aware it is un-enforceable as the bug bounty has to be paid by some entity. Good recommendation though.

Reviewer 3: Would it make sense to create a multisig arrangement tied to an address where people can donate for this purpose and bounties are paid out by multisig. I would personally donate 10 Ada per epoch to it.

Reviewer 4: What is meant by "unenforceable" and "paid by some entity"?

It feels like the whole bounty payout could be managed on-chain with the payout in ada or another native token (djed perhaps).

I haven't thought through this idea fully yet so be gentle...It feels like the whole flow could (maybe not should?) be managed on-chain.
"Responsible" party submits encrypted vulnerability details inside the metadata of [a] transaction(s).  Provides the decryption key to the "relevant parties" as well as payment address for any bounty (payment addr could be included in metadata). Once bug is resolved and paid the encryption key for the vuln data can be released so the public can view details

How much the bounty is worth will need to be determined somehow? Who votes on that? Going to be difficult without the public knowing the vuln details.  Feels like we need a separate "treasury pool" specifically for bug bounties that can be sined/paid by a subset of the governance members or a new class of dreps?

### #2: Emergency group to support the disaster recovery plan.
  -attacks on persons if known

>   -we don’t want attackers know who that group is?

Reviewer 3: ```suggestion
  -we don’t want attackers to know who that group is?
```

Recommendation:

### #3: Emergency change that needs to be done quickly.
  -a nuclear option
  -change requires “just do it” 
  -a security that fix must be implemented but CANNOT be communicated publicly

Recommendation:

### #4: Operational Security (OPSEC) of the committee.
  -physical security
  -digital security

Recommendation:

### #5: Committee keys lost or stolen or committee keys sold to a bad actor.

Recommendation:

> 

Reviewer 3: - Lost/Stolen : Member must be able to sign with the same wallet/stake key to authenticate for burn/mint of previous key

- Sold to bad actor : Loss of collateral / Keys destroyed by committee vote

Author: I think the result of lost, stolen, or sold is the same. The keys must be burned and new ones generated. I could be wrong.

### #6: DRep keys lost or stolen or DRep keys sold to a bad actor.

Recommendation:

> 

Author: I think the corrective action for lost, stolen, or sold is the same. The keys must be burned and new ones generated. Same as above. I think the difference in how new keys are generated is in the implementation.

### #7: Lack of governance for an extended period of time.
  -Uncontrolled hard forks by SPOs
  -Risk of no confidence

Recommendation:

### #8: Lack of trust in voter tooling.

Recommendation:

### #9: Trust in the wallets.
  -wallet voting attack vector
  -hardware wallet support

Recommendation:

### #10: Corruption - in any form, any number of actors.
  -Code of Conduct mitigation
  -on-chain mitigation

Recommendation:

### #11: Collusion - the bad version of collaboration to do something illegal or undesirable.

Recommendation:

> 

Reviewer 3: If this is collusion on the part of any committees members or DReps, it should result in loss of collateral by committee vote. Evidence needs to be presented in a reasonable and articulable way to a body (i.e. Oversight & Accountability Committee), with specific constitutional citations that explains the behavior and/or actions. Voting on such an action must be constitutionally based, not emotionally based.

Author: I don't know if the deposit to be a DRep can be revoked by any other entity. This could work. Will tag @JaredCorduan and @KtorZ to ask if this is technically possible or would require a code change?

Reviewer 5: as currently written, no, there is no way to revoke the deposit.

I'm personally not a fan of giving the CC this new power, which could  in theory could be abused.

Reviewer 3: I agree that the CC should have no powers beyond what is currently described in CIP-1694. I suppose I am envisioning a separate body to provide some oversight with issues like these.

### #12: Vote buying - Ada kickbacks.

Recommendation:

### #13: IGCO - initial governance coin offering
  -offering voters a token of unknown value by a representative to acquire delegators
  -including tokens other than Ada

Recommendation:

### #14: Powerful Voter Collusion
  -demotivates smaller bag holders
  -cause a loss of participation
  -too many voters abstain
  -minority controls the outcome
  -can push through any proposal

Recommendation:

### #15: Persistence - Removal of proposals by attackers.

Recommendation:

> 

Reviewer 3: I know I continue to suggest collateral. I'm going to do it here, too. If an individual or group of individuals want to remove a proposal, collateral should be required, just as submitting the proposal should. If the removal request is found by committee to be nefarious and/or without merit, a loss of collateral should occur.

Author: Same as above, I don't know if the deposit by the DRep is considered a collateral, or a deposit that can be only returned to the depositer. For example no other keyholder can revoke a stake pool deposit of 500 Ada. Only the keyholder can do it afaik.

Reviewer 4: I agree, collateral should be required.

### #16: Risk of Proposal overload - Risk of DDOS.

Recommendation:

> 

Reviewer 3: - Require collateral to accompany a proposal (i.e. 50 $Ada or higher)

- Collateral reimbursement once proposal is approved for a voting round or rejected due to constitutional violation

Author: Same as above, I think we need to define the difference between a deposit and collateral, and which entity would have the power to slash it. Also might need to consult the game theory experts on slashing mechanisms as a form of punishment.

### #17: Excessive actions.
  -Drawing too much money
  -Changing parameters that break the chain (block size=0 for example)
  -overlapping actions and duplicate actions

Recommendation:

> 

Reviewer 3: - Intentional attempts to submit parameter change proposals that are deemed detrimental to the health of the network should result in a loss of collateral. As mentioned for line 118, collateral should be required to submit proposals.
- Overlapping/duplicate actions should be mostly curbed due to collateral requirement. If 1:1 duplicate is successfully proposed, the enforcement of a penalty from collateral as a percentage might be appropriate.

Reviewer 4: Not sure there is a risk here.  We would need all governing bodies to agree that the submitted proposal is 1) valid and 2) everyone votes to pass it.

### #18: Lack of technical understanding by DReps.
Education of correct procedures for DReps

Recommendation:

> 

Reviewer 3: Curriculum framed from the contents of the constitution and the approved governance CIP (CIP-1694 or otherwise)

Author: I concur.

### #19: Sybil behavior - one person acting as multiple DReps.

Recommendation:

> 

Reviewer 3: Loss of collateral and expungement of registration. DReps should also be required to submit collateral with their registration. Not dissimilar from a minimum pledge with a stake pool.

Author: same as above, if there is a difference between collateral and deposit.

Reviewer 4: If the voting power is based on total ADA behind the DRep then what point would there be to create multiple registrations? To bloat the the chain? You could argue the transaction cost to register as a DRep would make this unattractive in the same way spamming the network with tiny transactions prevents DDoS

### #20: Overpower Risk - One DRep or entity with too much power.

Recommendation:

### #21: Superstar attack - one “superstar” person that everybody delegates to becomes a dictator with little skin in the game.
(Example: Elon Musk as a DRep)

> 

Reviewer 3: Collateral exceptions or tiers - A potential process during DRep registration where an approval/finalization process occurs by committee. A tiered collateral model might be appropriate, levied based on certain criteria (i.e. social influence, history of nefarious behavior)

Reviewer 4: Must be solved in the constitution (separation of powers again)

### #22: High chain load impacting governance.

Recommendation:

> 

Reviewer 3: - Run all governance actions on a sidechain
- gAda (Governance Ada) at 1:1 Mainnet Ada
- Snapshot based (every epoch)

Author: This sounds like a good recommendation like for example Project Catalyst uses a snapshot and the Jormungandr chain.

Reviewer 4: iirc a layer2 can share the same security guarantees as a layer1 but a sidechain is just a bridged, separate network -- offloading to layer2 definitely sounds like a good idea

### #23: Unknown external regulation risks - affecting a specific DRep by country or affecting governance as a whole.

Recommendation:

> 

Reviewer 3: - Form a Regulation Preparedness & Action Committee
- All issues of importance can be brought to the floor of the committee for evaluation, discussion, recommendation and/or action

Author: Maybe I need to break this into 2 types of risks.
1. Indvidual DRep.
2. Cardano Governance process as a whole.

1. I was thinking of just making it the responsibility of each DRep to comply with their local laws. Otherwise, legal fees would skyrocket. If a DRep from a particular region runs into legal issues with acting as a DRep they can either fight it or resign.

2. For Cardano governance process having legal issues, kick that over to the Cardano Foundation to fight.

What do you think?

Reviewer 3: Yes, I agree. Engaging with the CF for Governance legal challenges is probably the best course of action. At least initially, to get things aligned. It may be worth having a point of contact identified at CF as a resource for individual DReps who face legal challenges. If for nothing but some general guidance. Just having someone to talk to who understands the situation can be a huge help just for the stress factor alone.

### #24: Blocking a Constitutional Committee revote after a vote of “no confidence”.

Recommendation:

> 

Reviewer 3: - Proposal should be returned to the DReps with documentation that outlines why a vote of "no confidence" was given, giving the DRep body an opportunity to submit reasons and potential solutions to the proposer(s)
- or a direct revote by the DRep body requiring a strong percentage (3/4?) of votes to the affirmative in order to successfully override the "no confidence" vote.

### #25: Ensuring the legitimacy of side chains involved in governance.

Recommendation:

> 

Reviewer 3: - gNodes (Governance Nodes) implemented by process of election
- Disallow DReps and other committee members from operating gNodes
- Conflicts of interest should be outlined (i.e. spouses of gNode operators cannot be DReps and vice versa)
- Risk of collateral/pledge loss for circumventing outlined accepted practices
- Outline requirements for operating a gNode

Author: More nodes may not be desirable. Actions should be largely enforceable on-chain where or when possible. Unless the purpose of the nodes is the run the previously mentioned side chain like a jormungandr chain? Maybe leverage or recycle the Project Catalyst infrastructure?

Another option is to firewall out the sidechains from governance. That way people don't spin up a sidechain with a token of unknown value to capture the voting process on the main chain.

I don't know how accurate this link is, but it is an example of governance capture from an incentive external to the chain. Could be just the author's opinion: https://reddit.com/r/tezos/comments/12ahgl1/unaligned_governance_capture_is_it_a_risk_for/

## Use Cases

Use cases need to be fleshed out. Need more inputs here as each use case would apply to each prolem listed above. Seems like both the CIP and CPS formats is not a good fit for this document.

## Goals

The intent of this CPS is to promote awareness of governance security vulnerabilities and possible mitigation techniques in the Cardano ecosystem. The goal is to provide a comprehensive guide describing all known governance security threats and possible mitigiations techniques, but not to address specific software vulnerabilities. The governance security need is present regardless of the final form of CIP-1694 or any other governance document to minimize malicious activities that may occur within a system of voting, holding elections, and/or on-chain approval of automated actions are present.

## Open Questions

> TBD

Reviewer: ```suggestion
Proposed solutions should strive to address the questions outlined within [Problem](#problem).
```

---

> CPS: ?

Reviewer 6: ```suggestion
CPS: 6
```

Title: Cardano Governance Security

> Category: Meta

Reviewer 6: ```suggestion
Category: Ledger
```

In rare instances, a threat may be identified that requires responsible reporting procedures to prevent exposing an active vulnerability to bad actors seeking to exploit it. In such cases, it is essential to notify the appropriate group or entity with the ressources, or are responsible for mitigating the threat in a confidential manner.

### #1: Unknown software vulnerability affecting governance is discovered.

An assumption is that are no known software back doors. Unknown software vulnerabilities may be discovered.

**Recommendation:Pre-planned Response** The person discovering the vulnerability should take the following steps:
* Document the vulnerability: The first step is to document the vulnerability as accurately as possible. This documentation should include a detailed description of the vulnerability, steps to reproduce it, and any other relevant information.

> * Notify the relevant parties: The next step is to notify the appropriate parties, including the developers of the software, the blockchain community, and any relevant authorities. It's important to provide as much information as possible, so that the parties involved can understand the scope of the vulnerability and take the necessary steps to address it.

Reviewer 4: The "relevant parties" should (must?) be voted on by the community.  The contact detail data could (must?) be stored on chain.

8. Regularly review and update the security measures in place to ensure that they remain effective. Possibly do a periodic practice disaster recovery drill on one of the test networks.

### #4: Operational Security (OPSEC) of the committee.
  * Physical security
  * Digital security

**Recommendation:**

### #5: Committee keys lost or stolen or committee keys sold to a bad actor.

**Recommendation:**

### #6: DRep keys lost or stolen or DRep keys sold to a bad actor.

**Recommendation:**

### #7: Lack of governance for an extended period of time.
  * Uncontrolled hard forks by SPOs
  * Risk of no confidence

**Recommendation:**

### #8: Lack of trust in voter tooling.

**Recommendation:**

### #9: Trust in the wallets.
  * wallet voting attack vector
  * hardware wallet support

**Recommendation:**

### #10: Corruption - in any form, any number of actors.
  * Code of Conduct mitigation
  * on-chain mitigation

**Recommendation:**

### #11: Collusion - the bad version of collaboration to do something illegal or undesirable.

**Recommendation:**

### #12: Vote buying - Ada kickbacks.

**Recommendation:**

> 

Reviewer 4: As previously suggested in Item 11(Collusion) perhaps an on-chain naughty list that can be integrated into voting tools?.

but also who will police this? we'll need "courts" to determine guilt and folks to prosecute/defend

### #13: IGCO - initial governance coin offering
  * offering voters a token of unknown value by a representative to acquire delegators
  * including tokens other than Ada

> 

Reviewer 4: This feels like a duplicate of #12 (vote buying)

**Recommendation:**

### #14: Powerful Voter Collusion
  * demotivates smaller bag holders
  * cause a loss of participation
  * too many voters abstain
  * minority controls the outcome
  * can push through any proposal

**Recommendation:**

> 

Reviewer 4: Has to be addressed in the constitution.  Just like it is in the US constitution: https://constitution.congress.gov/browse/essay/intro-2-2-2/ALDE_00000031/ (I know nothing of this at any kind of depth but the idea is sound)

Recommendation:

### #16: Risk of Proposal overload - Risk of DDOS.

**Recommendation:**

### #17: Excessive actions.
  * Drawing too much money
  * Changing parameters that break the chain (block size=0 for example)
  * overlapping actions and duplicate actions

**Recommendation:**

### #18: Lack of technical understanding by DReps.
Education of correct procedures for DReps

**Recommendation:**

### #19: Sybil behavior - one person acting as multiple DReps.

**Recommendation:**

### #20: Overpower Risk - One DRep or entity with too much power.

**Recommendation:**

### #21: Superstar attack - one “superstar” person that everybody delegates to becomes a dictator with little skin in the game.
(Example: Elon Musk as a DRep)

### #22: High chain load impacting governance.

**Recommendation:**

### #23: Unknown external regulation risks - affecting a specific DRep by country or affecting governance as a whole.

**Recommendation:**

### #24: Blocking a Constitutional Committee revote after a vote of “no confidence”.

**Recommendation:**

### #25: Ensuring the legitimacy of side chains involved in governance.

**Recommendation:**

> 

Reviewer 4: need more data/description
